group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1735,6 +1756,10 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, CoordinatorRecord>
         final long currentTimeMs = time.milliseconds();
         final List<CoordinatorRecord> records = new ArrayList<>();
 
+        if (subscribedTopicRegex != null) {
+            groupRegexManager.validateAndRequestEval(groupId, subscribedTopicRegex);
+        }
```

Comment:
This does not seem to be the right place to schedule the resolution. I think that we should rather do it after we have done the main validation (e.g. member id, member epoch, etc.). A better may be in `hasMemberSubscriptionChanged`.

In `hasMemberSubscriptionChanged`, we could also be a bit more cleaver on the condition to trigger the resolution. For instance, we should check if the regex has been updated. If not, no point of scheduling the resolution now. We could also check if we already now the regular expression. If we do, no point of scheduling it too. If the regex has been updated, we also need to check whether the member was the last one using it. If it was, we need to generate a tombstone to delete it. For those, we may need to introduce a snapshottable HashMap<String, Integer> to keep a count of members per regex.

I also wonder where we are going to re-schedule the resolution of all the regexes after they expire. Do you already have an idea?